File: order-total-wrong-after-line-cancel-with-voucher/node/check-order-total-after-cancel.js
/**
 * Detect PrestaShop orders whose total went stale after a product line was
 * cancelled while a voucher was attached.
 *
 * Cancelling a product from an order in Back Office > Orders (OrderController /
 * the Order class) recalculates the remaining order_detail line totals, but it
 * does not re-derive total_discounts from the cart rules still attached to the
 * order (order_cart_rules). A cart rule computed as a percent-of-total, a fixed
 * amount, or free shipping was calculated once against the cart as it stood at
 * checkout, so once a line is cancelled that original cart total no longer
 * exists and the stored discount goes stale, along with total_paid and
 * total_paid_tax_incl. Tracked upstream across PrestaShop/PrestaShop issues
 * #17347, #23358, #23038, #28134, with the invalid-discount shape (negative or
 * tax_excl greater than tax_incl) tracked separately as issue #11059.
 *
 * This script defaults to detect and report only, since an order's total may
 * already be referenced by an invoice or an accounting export. The corrective
 * PUT to orders only runs under an explicit DRY_RUN=false override, always
 * prints a before/after diff, and never touches current_state (order state
 * changes belong only to POST /api/order_histories).
 *
 * Guide: https://www.allanninal.dev/prestashop/order-total-wrong-after-line-cancel-with-voucher/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const ORDER_IDS = (process.env.ORDER_IDS || "").split(",").map((s) => s.trim()).filter(Boolean);
const TOLERANCE = 0.02;

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

/**
 * Pure decision function, no I/O.
 *
 * orderDetails is the array of order_details rows still present on the order
 * (already fetched by the caller). orderCartRules is the array of
 * order_cart_rules rows for the order (already fetched). totalShipping and
 * reportedTotalTaxIncl are plain values already read from the order. Returns
 * the expected total, the delta against what the order reports, whether that
 * delta exceeds tolerance, and whether the cart rule values have the invalid
 * shape from issue #11059 (negative, or tax_excl sum greater than the
 * tax_incl sum).
 */
export function recomputeOrderTotal(orderDetails, orderCartRules, totalShipping,
                                     reportedTotalTaxIncl, tolerance = TOLERANCE) {
  const linesSum = orderDetails.reduce((sum, d) => sum + Number(d.total_price_tax_incl), 0);
  const activeRules = orderCartRules.filter((r) => String(r.deleted ?? "0") === "0");
  const cartRulesSum = activeRules.reduce((sum, r) => sum + Number(r.value), 0);
  const expected = linesSum + Number(totalShipping) - cartRulesSum;
  const reported = Number(reportedTotalTaxIncl);
  const delta = reported - expected;

  let invalidShape = activeRules.some((r) => Number(r.value) < 0);
  const taxExclSum = activeRules.reduce(
    (sum, r) => sum + Number(r.value_tax_excl ?? r.value),
    0
  );
  if (taxExclSum > cartRulesSum) invalidShape = true;

  return {
    expected_total: expected,
    reported_total: reported,
    delta,
    is_mismatched: Math.abs(delta) > tolerance,
    invalid_discount_shape: invalidShape,
  };
}

export function buildReportRow(idOrder, result, activeRuleIds) {
  return {
    id_order: idOrder,
    expected_total: Math.round(result.expected_total * 100) / 100,
    reported_total: Math.round(result.reported_total * 100) / 100,
    delta: Math.round(result.delta * 100) / 100,
    is_mismatched: result.is_mismatched,
    invalid_discount_shape: result.invalid_discount_shape,
    order_cart_rules_summed: activeRuleIds,
  };
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function getOrder(idOrder) {
  const data = await apiGet(`orders/${idOrder}`);
  return data.order || {};
}

async function orderDetailsFor(idOrder) {
  const data = await apiGet("order_details", {
    "filter[id_order]": idOrder,
    display: "full",
  });
  return data.order_details || [];
}

async function orderCartRulesFor(idOrder) {
  const data = await apiGet("order_cart_rules", {
    "filter[id_order]": idOrder,
    display: "full",
  });
  return data.order_cart_rules || [];
}

/**
 * Only called when DRY_RUN is explicitly false. Sends the full order body
 * back with corrected discount and paid totals. Never touches current_state;
 * state changes go only through POST /api/order_histories.
 */
async function applyCorrection(idOrder, order, result, activeRules) {
  const cartRulesSum = activeRules.reduce((sum, r) => sum + Number(r.value), 0);
  const taxExclSum = activeRules.reduce((sum, r) => sum + Number(r.value_tax_excl ?? r.value), 0);
  const corrected = { ...order };
  corrected.total_discounts = String(cartRulesSum);
  corrected.total_discounts_tax_incl = String(cartRulesSum);
  corrected.total_discounts_tax_excl = String(taxExclSum);
  corrected.total_paid = String(result.expected_total);
  corrected.total_paid_tax_incl = String(result.expected_total);
  corrected.total_paid_tax_excl = String(result.expected_total - (cartRulesSum - taxExclSum));
  delete corrected.current_state;

  console.warn(`BEFORE: total_paid_tax_incl=${order.total_paid_tax_incl} total_discounts=${order.total_discounts}`);
  console.warn(`AFTER:  total_paid_tax_incl=${corrected.total_paid_tax_incl} total_discounts=${corrected.total_discounts}`);

  const url = new URL(`${PRESTASHOP_URL}/api/orders/${idOrder}`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ order: corrected }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT orders/${idOrder}`);
  return res.json();
}

export async function run() {
  if (!ORDER_IDS.length) {
    console.error("Set ORDER_IDS to a comma separated list of order ids to check.");
    return;
  }

  let flagged = 0;
  for (const idOrder of ORDER_IDS) {
    const order = await getOrder(idOrder);
    if (!order || !Object.keys(order).length) {
      console.warn(`Order ${idOrder} not found, skipping.`);
      continue;
    }
    const details = await orderDetailsFor(idOrder);
    const cartRules = await orderCartRulesFor(idOrder);
    const activeRules = cartRules.filter((r) => String(r.deleted ?? "0") === "0");

    const result = recomputeOrderTotal(
      details, cartRules, order.total_shipping || "0", order.total_paid_tax_incl || "0"
    );
    if (!(result.is_mismatched || result.invalid_discount_shape)) continue;

    const row = buildReportRow(idOrder, result, activeRules.map((r) => r.id));
    flagged++;
    console.warn(
      `Order ${row.id_order} total mismatch. expected=${row.expected_total.toFixed(2)} ` +
        `reported=${row.reported_total.toFixed(2)} delta=${row.delta.toFixed(2)} ` +
        `invalid_discount_shape=${row.invalid_discount_shape} cart_rules=${JSON.stringify(row.order_cart_rules_summed)}`
    );

    if (!DRY_RUN) {
      await applyCorrection(idOrder, order, result, activeRules);
      console.log(`Order ${idOrder} corrected via PUT /api/orders/${idOrder}.`);
    }
  }

  console.log(
    `Done. ${flagged} order(s) flagged. DRY_RUN=${DRY_RUN} (repair only runs when explicitly false).`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: order-total-wrong-after-line-cancel-with-voucher/python/check_order_total_after_cancel.py
"""Detect PrestaShop orders whose total went stale after a product line was
cancelled while a voucher was attached.

Cancelling a product from an order in Back Office > Orders (OrderController /
the Order class) recalculates the remaining order_detail line totals, but it
does not re-derive total_discounts from the cart rules still attached to the
order (order_cart_rules). A cart rule computed as a percent-of-total, a fixed
amount, or free shipping was calculated once against the cart as it stood at
checkout, so once a line is cancelled that original cart total no longer
exists and the stored discount goes stale, along with total_paid and
total_paid_tax_incl. Tracked upstream across PrestaShop/PrestaShop issues
#17347, #23358, #23038, #28134, with the invalid-discount shape (negative or
tax_excl greater than tax_incl) tracked separately as issue #11059.

This script defaults to detect and report only, since an order's total may
already be referenced by an invoice or an accounting export. The corrective
PUT to orders only runs under an explicit DRY_RUN=false override, always
prints a before/after diff, and never touches current_state (order state
changes belong only to POST /api/order_histories).

Run on a schedule for orders touched since the last run. Safe to run again
and again in report mode.
"""
import os
import logging
from decimal import Decimal

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_order_total_after_cancel")

PRESTASHOP_URL = os.environ.get("PRESTASHOP_URL", "https://demo.example.com").rstrip("/")
PRESTASHOP_WS_KEY = os.environ.get("PRESTASHOP_WS_KEY", "WSKEYDUMMY")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
ORDER_IDS = [o.strip() for o in os.environ.get("ORDER_IDS", "").split(",") if o.strip()]
AUTH = (PRESTASHOP_WS_KEY, "")

TOLERANCE = Decimal("0.02")


def recompute_order_total(order_details, order_cart_rules, total_shipping,
                           reported_total_tax_incl, tolerance=TOLERANCE):
    """Pure decision function, no I/O.

    order_details is the list of order_details rows still present on the
    order (already fetched by the caller). order_cart_rules is the list of
    order_cart_rules rows for the order (already fetched). total_shipping and
    reported_total_tax_incl are plain values already read from the order.
    Returns the expected total, the delta against what the order reports,
    whether that delta exceeds tolerance, and whether the cart rule values
    have the invalid shape from issue #11059 (negative, or tax_excl sum
    greater than the tax_incl sum).
    """
    lines_sum = sum(
        (Decimal(str(d["total_price_tax_incl"])) for d in order_details),
        Decimal("0"),
    )
    active_rules = [r for r in order_cart_rules if str(r.get("deleted", "0")) == "0"]
    cart_rules_sum = sum(
        (Decimal(str(r["value"])) for r in active_rules),
        Decimal("0"),
    )
    expected = lines_sum + Decimal(str(total_shipping)) - cart_rules_sum
    reported = Decimal(str(reported_total_tax_incl))
    delta = reported - expected

    invalid_shape = any(Decimal(str(r["value"])) < 0 for r in active_rules)
    tax_excl_sum = sum(
        (Decimal(str(r.get("value_tax_excl", r["value"]))) for r in active_rules),
        Decimal("0"),
    )
    if tax_excl_sum > cart_rules_sum:
        invalid_shape = True

    return {
        "expected_total": expected,
        "reported_total": reported,
        "delta": delta,
        "is_mismatched": abs(delta) > tolerance,
        "invalid_discount_shape": invalid_shape,
    }


def build_report_row(id_order, result, active_rule_ids):
    return {
        "id_order": id_order,
        "expected_total": round(float(result["expected_total"]), 2),
        "reported_total": round(float(result["reported_total"]), 2),
        "delta": round(float(result["delta"]), 2),
        "is_mismatched": result["is_mismatched"],
        "invalid_discount_shape": result["invalid_discount_shape"],
        "order_cart_rules_summed": active_rule_ids,
    }


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def get_order(id_order):
    data = api_get(f"orders/{id_order}")
    return data.get("order") or {}


def order_details_for(id_order):
    data = api_get("order_details", params={
        "filter[id_order]": id_order,
        "display": "full",
    })
    return data.get("order_details") or []


def order_cart_rules_for(id_order):
    data = api_get("order_cart_rules", params={
        "filter[id_order]": id_order,
        "display": "full",
    })
    return data.get("order_cart_rules") or []


def apply_correction(id_order, order, result, active_rules):
    """Only called when DRY_RUN is explicitly false. Sends the full order
    body back with corrected discount and paid totals. Never touches
    current_state; state changes go only through POST /api/order_histories.
    """
    corrected = dict(order)
    cart_rules_sum = sum(Decimal(str(r["value"])) for r in active_rules) if active_rules else Decimal("0")
    tax_excl_sum = sum(
        Decimal(str(r.get("value_tax_excl", r["value"]))) for r in active_rules
    ) if active_rules else Decimal("0")
    corrected["total_discounts"] = str(cart_rules_sum)
    corrected["total_discounts_tax_incl"] = str(cart_rules_sum)
    corrected["total_discounts_tax_excl"] = str(tax_excl_sum)
    corrected["total_paid"] = str(result["expected_total"])
    corrected["total_paid_tax_incl"] = str(result["expected_total"])
    corrected["total_paid_tax_excl"] = str(result["expected_total"] - (cart_rules_sum - tax_excl_sum))
    corrected.pop("current_state", None)

    log.warning("BEFORE: total_paid_tax_incl=%s total_discounts=%s",
                order.get("total_paid_tax_incl"), order.get("total_discounts"))
    log.warning("AFTER:  total_paid_tax_incl=%s total_discounts=%s",
                corrected["total_paid_tax_incl"], corrected["total_discounts"])

    r = requests.put(
        f"{PRESTASHOP_URL}/api/orders/{id_order}",
        params={"output_format": "JSON"},
        json={"order": corrected},
        auth=AUTH,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    if not ORDER_IDS:
        log.error("Set ORDER_IDS to a comma separated list of order ids to check.")
        return

    flagged = 0
    for id_order in ORDER_IDS:
        order = get_order(id_order)
        if not order:
            log.warning("Order %s not found, skipping.", id_order)
            continue
        details = order_details_for(id_order)
        cart_rules = order_cart_rules_for(id_order)
        active_rules = [r for r in cart_rules if str(r.get("deleted", "0")) == "0"]

        result = recompute_order_total(
            details, cart_rules, order.get("total_shipping", "0"),
            order.get("total_paid_tax_incl", "0"),
        )
        if not (result["is_mismatched"] or result["invalid_discount_shape"]):
            continue

        row = build_report_row(id_order, result, [r.get("id") for r in active_rules])
        flagged += 1
        log.warning(
            "Order %s total mismatch. expected=%.2f reported=%.2f delta=%.2f "
            "invalid_discount_shape=%s cart_rules=%s",
            row["id_order"], row["expected_total"], row["reported_total"],
            row["delta"], row["invalid_discount_shape"], row["order_cart_rules_summed"],
        )

        if not DRY_RUN:
            apply_correction(id_order, order, result, active_rules)
            log.info("Order %s corrected via PUT /api/orders/%s.", id_order, id_order)

    log.info(
        "Done. %d order(s) flagged. DRY_RUN=%s (repair only runs when explicitly false).",
        flagged, DRY_RUN,
    )


if __name__ == "__main__":
    run()
